server/plugins/LFUGens.cpp

```diff
@@ -793,109 +799,281 @@ void LFGauss_Ctor(LFGauss* unit) {
+        ZXP(out) = z;
 
-    unit->mPhase = phase - phaseOffset;
-    unit->mPhaseOffset = phaseOffset;
-}
+        phase += phaseInc; prev_phaseInc = phaseInc;);
 
+    unit->mPhase = phase;
+    unit->mPhaseOffset = phaseOff;
+    unit->mPhaseIncrement = prev_phaseInc;
+}
 
-void Impulse_next_k(Impulse* unit, int inNumSamples) {
+void Impulse_next_aa(Impulse* unit, int inNumSamples) {
     float* out = ZOUT(0);
-    float freq = ZIN0(0) * unit->mFreqMul;
+    double phase = unit->mPhase;
+
+    double phaseInc = unit->mPhaseIncrement;
+    float* freqin = ZIN(0);
+    float freqmul = unit->mFreqMul;
+
+    double prev_phaseOff = unit->mPhaseOffset;
+    float* phaseOffIn = ZIN(1);
+
+    LOOP1(inNumSamples, float z = Impulse_testWrapPhase(phaseInc, phase); float phaseOff = ZXP(phaseOffIn);
+          float phaseOffInc = phaseOff - prev_phaseOff; phase += phaseOffInc; Impulse_testWrapPhase(phaseInc, phase);
+          phaseInc = ZXP(freqin) * freqmul; ZXP(out) = z;
+
+          phase += phaseInc; prev_phaseOff = phaseOff;);
 
+    unit->mPhase = phase;
+    unit->mPhaseOffset = prev_phaseOff;
+    unit->mPhaseIncrement = phaseInc;
+}
+
+void Impulse_next_ai(Impulse* unit, int inNumSamples) {
+    float* out = ZOUT(0);
     double phase = unit->mPhase;
-    LOOP1(
-        inNumSamples, float z; if (phase >= 1.f) {
-            phase -= 1.f;
-            z = 1.f;
-        } else { z = 0.f; } phase += freq;
-        ZXP(out) = z;);
+
+    double phaseInc = unit->mPhaseIncrement;
+    float* freqin = ZIN(0);
+    float freqmul = unit->mFreqMul;
+
+    LOOP1(inNumSamples, float z = Impulse_testWrapPhase(phaseInc, phase); phaseInc = ZXP(freqin) * freqmul;
+          ZXP(out) = z; phase += phaseInc;);
 
     unit->mPhase = phase;
+    unit->mPhaseIncrement = phaseInc;
 }
 
+// Note: In calc functions,
+// phase offset change is added to phase, then phase is wrapped into range,
+// followed by phase increment (freq), then tested for triggering an impulse.
+// So a freq increment will trigger an impulse, but not a phase offset.
 void Impulse_Ctor(Impulse* unit) {
-    unit->mPhase = ZIN0(1);
+    unit->mPhaseOffset = ZIN0(1);
+    unit->mFreqMul = unit->mRate->mSampleDur;
+    unit->mPhaseIncrement = ZIN0(0) * unit->mFreqMul;
 
-    if (INRATE(0) == calc_FullRate) {
-        if (INRATE(1) != calc_ScalarRate) {
-            SETCALC(Impulse_next_ak);
-            unit->mPhase = 1.f;
-        } else {
-            SETCALC(Impulse_next_a);
+    double initphaseOff = unit->mPhaseOffset;
+    double initPhaseInc = unit->mPhaseIncrement;
+    double initPhase = sc_wrap(initphaseOff, 0.0, 1.0);
+
+    // Initial phase offset of 0 means output of 1 on first sample.
+    // Set phase to wrap point to trigger impulse on first sample
+    if (initPhase == 0.0 && initPhaseInc >= 0.0) {
+        initPhase = 1.0; // positive frequency trigger/wrap position
+    }
+    unit->mPhase = initPhase;
+
+    UnitCalcFunc func;
+    switch (INRATE(0)) {
```

Comment:
I don't know how to address it, but when this reaches the stage of c++ style review, I expect there to be a comment about the number of calculation functions. There have been some discussions lately about the topic of handling the permutations.

If you remove all the `i` functions, then we go from 9 down to 4 -- reducing the maintenance load by a factor of 2.22, while the performance gain of the `i` functions would be minuscule. So I'd start there.

Reply:
Yes, 9 is overkill. I've trimmed it to 7 (dropping a-rate phase offset for k- and i-rate frequencies). If needed we could also drop `a`-rate phase offset altogether. I kept it because I recall a user on the scsynth forums being surprised it wasn't supported.
I agree we could drop `i`-rate altogether, falling back on a `k`-rate minimum for both args. Although one of the goals of this PR was support more rates. If maintenance is the issue though, i-rate is the easiest to introspect.